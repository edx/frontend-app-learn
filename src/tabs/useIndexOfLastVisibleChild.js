import { useLayoutEffect, useRef, useState } from 'react';
import useWindowSize from './useWindowSize';

const invisibleStyle = {
  position: 'absolute',
  left: 0,
  pointerEvents: 'none',
  visibility: 'hidden',
};

export default function useIndexOfLastVisibleChild() {
  const containerElementRef = useRef(null);
  const overflowElementRef = useRef(null);
  const containingRectRef = useRef({});
  const [indexOfLastVisibleChild, setIndexOfLastVisibleChild] = useState(-1);
  const windowSize = useWindowSize();

  useLayoutEffect(() => {
    const containingRect = containerElementRef.current.getBoundingClientRect();

    // No-op if the width is unchanged.
    // (Assumes tabs themselves don't change count or width).
    if (!containingRect.width === containingRectRef.current.width) {
      return;
    }
    // Update for future comparison
    containingRectRef.current = containingRect;

    // Get array of child nodes from NodeList form
    const childNodesArr = Array.prototype.slice.call(containerElementRef.current.children);
    const { nextIndexOfLastVisibleChild } = childNodesArr
      // filter out the overflow element
      .filter(childNode => childNode !== overflowElementRef.current)
      // sum the widths to find the last visible element's index
      .reduce((acc, childNode, index) => {
        // use floor to prevent rounding errors
        acc.sumWidth += Math.floor(childNode.getBoundingClientRect().width);
        if (acc.sumWidth <= containingRect.width) {
          acc.nextIndexOfLastVisibleChild = index;
        }
        return acc;
      }, {
        // Include the overflow element's width to begin with. Doing this means
        // sometimes we'll show a dropdown with one item in it when it would fit,
        // but allowing this case dramatically simplifies the calculations we need
        // to do above.
        sumWidth: overflowElementRef.current ? overflowElementRef.current.getBoundingClientRect().width : 0,
        nextIndexOfLastVisibleChild: -1,
      });


    setIndexOfLastVisibleChild(nextIndexOfLastVisibleChild);
  }, [windowSize, containerElementRef.current]);

  return [indexOfLastVisibleChild, containerElementRef, invisibleStyle, overflowElementRef];
}
