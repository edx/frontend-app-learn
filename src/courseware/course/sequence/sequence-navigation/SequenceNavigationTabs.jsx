import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import UnitButton from './UnitButton';
import SequenceNavigationDropdown from './SequenceNavigationDropdown';
import useIndexOfLastVisibleChild from '../../../../generic/tabs/useIndexOfLastVisibleChild';
import { useIsOnXLDesktop, useIsSidebarOpen } from './hooks';

const SequenceNavigationTabs = ({
  unitIds, unitId, showCompletion, onNavigate,
}) => {
  const isSidebarOpen = useIsSidebarOpen(unitId);
  const [
    indexOfLastVisibleChild,
    containerRef,
    invisibleStyle,
  ] = useIndexOfLastVisibleChild();
  const isOnXLDesktop = useIsOnXLDesktop();
  const shouldDisplayDropdown = indexOfLastVisibleChild === -1;

  return (
    <div style={{ flexBasis: '100%', minWidth: 0 }}>
      <div
        ref={containerRef}
        className={classNames('sequence-navigation-tabs-container', {
          'navigation-tab-width': isOnXLDesktop && isSidebarOpen,
        })}
      >
        <div
          className="sequence-navigation-tabs d-flex flex-grow-1"
          style={shouldDisplayDropdown ? invisibleStyle : null}
        >
          {unitIds.map(buttonUnitId => (
            <UnitButton
              key={buttonUnitId}
              unitId={buttonUnitId}
              isActive={unitId === buttonUnitId}
              showCompletion={showCompletion}
              onClick={onNavigate}
            />
          ))}
        </div>
      </div>
      {shouldDisplayDropdown && (
        <SequenceNavigationDropdown
          unitId={unitId}
          onNavigate={onNavigate}
          showCompletion={showCompletion}
          unitIds={unitIds}
        />
      )}
    </div>
  );
};

SequenceNavigationTabs.propTypes = {
  unitId: PropTypes.string.isRequired,
  onNavigate: PropTypes.func.isRequired,
  showCompletion: PropTypes.bool.isRequired,
  unitIds: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default SequenceNavigationTabs;
