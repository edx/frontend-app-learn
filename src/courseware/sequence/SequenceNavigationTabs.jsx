import React from 'react';
import PropTypes from 'prop-types';
import { Dropdown } from '@edx/paragon';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import UnitButton from './UnitButton';
import useIndexOfLastVisibleChild from '../../tabs/useIndexOfLastVisibleChild';

const invisibleStyle = {
  position: 'absolute',
  left: 0,
  pointerEvents: 'none',
  visibility: 'hidden',
};

export default function SequenceNavigationTabs({
  unitIds, activeUnitId, showCompletion, onNavigate,
}) {
  const [
    indexOfLastVisibleChild,
    containerRef,
  ] = useIndexOfLastVisibleChild();
  const shouldDisplayDropdown = indexOfLastVisibleChild === -1;

  return (
    <div style={{ flexBasis: '100%', minWidth: 0 }}>
      <div className="sequence-navigation-tabs-container" ref={containerRef}>
        <div
          className="sequence-navigation-tabs"
          style={shouldDisplayDropdown ? invisibleStyle : null}
        >
          {unitIds.map(unitId => (
            <UnitButton
              key={unitId}
              unitId={unitId}
              isActive={activeUnitId === unitId}
              showCompletion={showCompletion}
              onClick={onNavigate}
            />
          ))}

        </div>
      </div>
      {shouldDisplayDropdown && (
        <Dropdown className="sequence-navigation-dropdown">
          <Dropdown.Button className="dropdown-button font-weight-normal w-100 border-right-0">
            <FormattedMessage
              id="learn.course.sequence.navigation.mobile.menu"
              description="The title of the mobile menu for sequence navigation of units"
              defaultMessage="{current} of {total}"
              values={{
                current: unitIds.indexOf(activeUnitId) + 1,
                total: unitIds.length,
              }}
            />
          </Dropdown.Button>
          <Dropdown.Menu className="w-100">

            {unitIds.map(unitId => (
              <UnitButton
                className="w-100"
                key={unitId}
                unitId={unitId}
                isActive={activeUnitId === unitId}
                showCompletion={showCompletion}
                onClick={onNavigate}
                showTitle
              />
            ))}

          </Dropdown.Menu>
        </Dropdown>
      )}
    </div>
  );
}

SequenceNavigationTabs.propTypes = {
  activeUnitId: PropTypes.string.isRequired,
  onNavigate: PropTypes.func.isRequired,
  showCompletion: PropTypes.bool.isRequired,
  unitIds: PropTypes.arrayOf(PropTypes.string).isRequired,
};
