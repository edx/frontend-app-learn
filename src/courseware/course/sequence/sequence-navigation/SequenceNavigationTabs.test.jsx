import React from 'react';
import { render, screen } from '../../../../test/test-utils';
import SequenceNavigationTabs from './SequenceNavigationTabs';
import useIndexOfLastVisibleChild from '../../../../tabs/useIndexOfLastVisibleChild';

// Mock the hook to avoid relying on its implementation and mocking `getBoundingClientRect`.
jest.mock('../../../../tabs/useIndexOfLastVisibleChild');

describe('Sequence Navigation Tabs', () => {
  const testUnits = [...Array(10).keys()].map(i => String(i + 1));

  const initialState = {
    models: {
      units: testUnits.reduce(
        (acc, unitId) => Object.assign(acc, {
          [unitId]: {
            contentType: 'other',
            title: unitId,
          },
        }),
        {},
      ),
    },
  };

  const mockData = {
    unitId: '1',
    onNavigate: () => {
    },
    showCompletion: false,
    unitIds: testUnits,
  };

  it('renders correctly without dropdown', () => {
    useIndexOfLastVisibleChild.mockReturnValue([0, null, null]);
    const { asFragment } = render(<SequenceNavigationTabs {...mockData} />, { initialState });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly with dropdown', () => {
    useIndexOfLastVisibleChild.mockReturnValue([-1, null, null]);
    const { asFragment } = render(<SequenceNavigationTabs {...mockData} />, { initialState });
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders unit buttons', () => {
    useIndexOfLastVisibleChild.mockReturnValue([0, null, null]);
    render(<SequenceNavigationTabs {...mockData} />, { initialState });
    expect(screen.getAllByRole('button').length).toEqual(testUnits.length);
  });

  it('renders unit buttons and dropdown button', () => {
    useIndexOfLastVisibleChild.mockReturnValue([-1, null, null]);
    render(<SequenceNavigationTabs {...mockData} />, { initialState });
    expect(screen.getAllByRole('button').length).toEqual(testUnits.length + 1);
  });
});
