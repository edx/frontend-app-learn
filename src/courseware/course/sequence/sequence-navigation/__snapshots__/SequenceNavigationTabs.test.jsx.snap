// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sequence Navigation Tabs renders correctly with dropdown 1`] = `
<DocumentFragment>
  <div
    style="flex-basis: 100%; min-width: 0;"
  >
    <div
      class="sequence-navigation-tabs-container"
    >
      <div
        class="sequence-navigation-tabs d-flex flex-grow-1"
      >
        <button
          class="btn active"
          title="1"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="2"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="3"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="4"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="5"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="6"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="7"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="8"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="9"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="10"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
      </div>
    </div>
    <div
      class="dropdown sequence-navigation-dropdown"
    >
      <button
        aria-expanded="false"
        aria-haspopup="true"
        class="dropdown-toggle btn dropdown-button font-weight-normal w-100 border-right-0"
        id="pgn__dropdown-trigger-0"
      >
        <span>
          1 of 10
        </span>
      </button>
      <div
        aria-hidden="true"
        aria-labelledby="pgn__dropdown-trigger-0"
        class="dropdown-menu w-100"
        role="menu"
      >
        <button
          class="btn active w-100"
          title="1"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            1
          </span>
        </button>
        <button
          class="btn w-100"
          title="2"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            2
          </span>
        </button>
        <button
          class="btn w-100"
          title="3"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            3
          </span>
        </button>
        <button
          class="btn w-100"
          title="4"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            4
          </span>
        </button>
        <button
          class="btn w-100"
          title="5"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            5
          </span>
        </button>
        <button
          class="btn w-100"
          title="6"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            6
          </span>
        </button>
        <button
          class="btn w-100"
          title="7"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            7
          </span>
        </button>
        <button
          class="btn w-100"
          title="8"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            8
          </span>
        </button>
        <button
          class="btn w-100"
          title="9"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            9
          </span>
        </button>
        <button
          class="btn w-100"
          title="10"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
          <span
            class="unit-title"
          >
            10
          </span>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Sequence Navigation Tabs renders correctly without dropdown 1`] = `
<DocumentFragment>
  <div
    style="flex-basis: 100%; min-width: 0;"
  >
    <div
      class="sequence-navigation-tabs-container"
    >
      <div
        class="sequence-navigation-tabs d-flex flex-grow-1"
      >
        <button
          class="btn active"
          title="1"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="2"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="3"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="4"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="5"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="6"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="7"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="8"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="9"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
        <button
          class="btn"
          title="10"
          type="button"
        >
          <img
            alt="fa-book"
            class="fa-book"
            data-testid="icon"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
